2016/proposals/singh-shubham.md

```diff
@@ -0,0 +1,210 @@
+# Manage workflow for instructor training
+
+## Abstract
+
+Computing Skills have now become essential more so in the field of Science, but many a time we find researchers struggling with code or tools to get things done.
+Software Carpentary is a step in this direction.
+Since 1998, Software Carpentry has been teaching researchers in science, engineering, medicine, and related disciplines the computing skills they need to get more done in less time and with less pain.
+Software Carpentary , uses AMY ( a django based web application) to to keep track of who's qualified to teach, what workshops have run, who taught and attended. As more and more people are applying to become instructors at Software Carpentary. It has become essential to manage the workflow for instructor training.
+
+This project aims at designing such a workflow and implementing it by adding appropriate views to AMY.
+
+## Proposed workflow
+
+![HTML MOCKUP flowchart](https://cloud.githubusercontent.com/assets/15982349/13770094/aa23d670-eaaa-11e5-81a8-312f860ecd7b.png)
+
+1. Interested Trainees apply as Group or Individuals by filling forms.
+![HTML MOCKUP FORMS](https://cloud.githubusercontent.com/assets/15982349/13816694/6c7a68e6-ebb5-11e5-809e-eb5bd00d7a22.png)
+2. All applications are listed in the *ALL_APPLICATIONS* view. In this view the lead trainer can accept or reject an application.
+The lead trainer has to compose one sample email by filling a web-form before accepting an application, this email will contain information such as
+date, location, mode(online or in person), instructor name and instructor email id. This information is also stored in the database for the accepted applicant to be queried on later.
+Accepting an application will send an email, to the applicant that he/she has been accepted into the program with other details about the actual training along with one randomly generated password to login in to their account (this password can be changed to anything by the trainee after the first login).
+The sample email has to be written once and can be sent to all the people who have been accepted for a particular training session.
+For another training session the lead trainer would just have to fill the web-form again, overwriting previous values.
+![HTML MOCKUP ALL_APPLICATIONS]( https://cloud.githubusercontent.com/assets/15982349/13816688/66a3e7e4-ebb5-11e5-83f1-5086ddddac13.png)
+
+3. All accepted applications, are then transferred to the Trainee view. In this view the lead trainer, trainers,discussion leaders, lesson maintainers can track progress of the trainees, provide feedback and update the status of every stage. They are also notified here if the Trainee has requested a discussion session.
```

Comment:
This change can be incorporated into the workflow. Since we already know the staring date of the training session, we could keep a timer for a month, and as soon as we receive a feedback from the discussion  leaders, we could stop the timer , otherwise the training automatically fails.

A more elegant way to handle this, is by maintaining two more fields, which I hadn't done before. Discussion attempts and Homework submissions.
Discussion attempts would be an integer field, updated by the discussion leader confirming that the trainee has attended a particular discussion session. It would start from zero and would be incremented as the trainee attends more and more discussion sessions.
Homework submission - This would also be an integer field, but this would be updated by the trainee from the Trainee home view. The trainee would be required to paste the URL of the pull request made. We could run a validator confirming that this is indeed a repository of SWC or DC , which could be confirmed by simply analysing the URL, as it would be of the form - https://github.com/swcarpentry/amy/pull/(PR NO).
or 
we could use  [github api](https://developer.github.com/v3/)  to list down all the pull requests to the repository and check the if this PR actually exists and confirm that it was created by the trainee. ( this approach I think would probably be an overkill, as later on the lesson maintainers would be manually be checking the quality and the credibility of the above pull request ).